File: backend/app/schemas/garden_v2.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class GardenJourneyContextResponse(BaseModel):
    journey_day: int
    plot_index: int
    plot_day: int
    plot_size_days: int


class GardenPlotResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int

    plot_index: int
    start_journey_day: int
    end_journey_day: int

    title: str
    status: str

    rows: int
    columns: int

    created_at: datetime
    updated_at: datetime


class GardenObjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    garden_plot_id: int

    element_type: str
    object_type: str
    object_subtype: str

    source_type: str
    source_id: int | None

    position_row: int
    position_column: int
    layer: int

    status: str
    is_persistent: bool
    visible_date: date | None

    title: str
    description: str | None
    metadata_json: str | None

    created_at: datetime
    updated_at: datetime


class HabitTreeStateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    habit_id: int

    growth_points: int
    current_streak: int
    best_streak: int

    last_completed_date: date | None
    is_dormant: bool

    active_cycle_number: int

    created_at: datetime
    updated_at: datetime


class HabitTreeCycleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    habit_id: int
    garden_plot_id: int
    garden_object_id: int | None

    cycle_number: int
    cycle_start_growth_point: int
    cycle_end_growth_point: int
    growth_points_in_cycle: int

    status: str

    created_at: datetime
    updated_at: datetime


class GardenPlotDetailResponse(BaseModel):
    plot: GardenPlotResponse
    objects: list[GardenObjectResponse]


class GardenWorldResponse(BaseModel):
    context: GardenJourneyContextResponse
    plots: list[GardenPlotResponse]


class GardenCurrentPlotResponse(BaseModel):
    context: GardenJourneyContextResponse
    plot: GardenPlotResponse
    objects: list[GardenObjectResponse]


class GardenV2TaskSyncResponse(BaseModel):
    created_count: int
    skipped_count: int
    objects: list[GardenObjectResponse]


class GardenV2HabitTreeSyncResponse(BaseModel):
    changed_count: int
    skipped_count: int
    dormant_count: int
    objects: list[GardenObjectResponse]


class GardenV2WaterAreaSyncResponse(BaseModel):
    changed: bool
    completed_water_tasks_in_plot: int
    object: GardenObjectResponse | None

File: frontend/src/types/gardenV2.ts
export type GardenV2JourneyContext = {
  journey_day: number;
  plot_index: number;
  plot_day: number;
  plot_size_days: number;
};

export type GardenV2Plot = {
  id: number;
  user_id: number;

  plot_index: number;
  start_journey_day: number;
  end_journey_day: number;

  title: string;
  status: string;

  rows: number;
  columns: number;

  created_at: string;
  updated_at: string;
};

export type GardenV2Object = {
  id: number;
  user_id: number;
  garden_plot_id: number;

  element_type: string;
  object_type: string;
  object_subtype: string;

  source_type: string;
  source_id: number | null;

  position_row: number;
  position_column: number;
  layer: number;

  status: string;
  is_persistent: boolean;
  visible_date: string | null;

  title: string;
  description: string | null;
  metadata_json: string | null;

  created_at: string;
  updated_at: string;
};

export type GardenV2CurrentPlotResponse = {
  context: GardenV2JourneyContext;
  plot: GardenV2Plot;
  objects: GardenV2Object[];
};

export type GardenV2AirReward = {
  user_air_reward_id: number;
  task_id: number;
  journey_day: number;
  reward_date: string;
  awarded_at: string;

  reward_code: string;
  reward_name: string;
  reward_description: string | null;
  object_subtype: string;

  garden_object: GardenV2Object | null;
};